### ISSUE TITLE
Docker/Image Cleanup 

### ISSUE THREAD

**Issue by TheChymera** [IBT-FMI/gebuilder#17]
It seems that going forward the last issue in the way of turning this into a semiautomated pilot is having some sort of automated cleanup process (even if it's not a very smart one).

This was easy for [image file cleanup](https://github.com/IBT-FMI/gebuilder/blob/0fe275705b2a0ca8b8ac8538b0be87339a7ba3f1/gebuilder/example_hooks/openstack_image/95-cleanup.sh).

However Docker seems to have [serious issues](https://github.com/moby/moby/issues/3182) when using devicemapper.
Currently I have put together [this abomination](https://github.com/IBT-FMI/gebuilder/blob/be4632ef7898984000ae1dd213e1e7cf82444a2e/gebuilder/example_hooks/docker_image/post/95-cleanup.sh), which kind-of works in wiping **all** Docker images (suboptimal, but all things considered acceptable solution) given a reboot. Needless to say that's in no way ok.

@Doeme thoughts?

**Comment by Doeme** [comment]
Yes. This code is pure art (especially lines 17-19 and 27-28). I mean, not like the pleasant one, rather like [Pendereckis Threnody for the Victims of Hiroshima](http://www.youtube.com/watch?v=Dp3BlFZWJNA) or the whole Discography of [Gulaggh/Stalaggh](http://www.youtube.com/watch?v=nuO9xtNsRs0).

Otherwise, have you tried using some of the suggestions from here:
https://stackoverflow.com/questions/27853571/why-is-docker-image-eating-up-my-disk-space-that-is-not-used-by-docker
e.g.

* `docker system prune --force`
* `docker ps -qa | xargs docker inspect --format='{{ .State.Pid }}' | xargs -IZ fstrim /proc/Z/root/`
* ...

**Comment by TheChymera** [comment]
I have tried that, and much more, just a sample:

```
builder /var/lib/docker # du -cksh
8.8G    .
8.8G    total
builder /var/lib/docker # docker system prune --force
Total reclaimed space: 0B
builder /var/lib/docker # docker ps -qa | xargs docker inspect --format='{{ .State.Pid }}' | xargs -IZ fstrim /proc/Z/root/
"docker inspect" requires at least 1 argument.
See 'docker inspect --help'.

Usage:  docker inspect [OPTIONS] NAME|ID [NAME|ID...]

Return low-level information on Docker objects
builder /var/lib/docker # docker rm -v $(docker ps -a -q -f status=exited)
"docker rm" requires at least 1 argument.
See 'docker rm --help'.

Usage:  docker rm [OPTIONS] CONTAINER [CONTAINER...]

Remove one or more containers
builder /var/lib/docker # docker rmi $(docker images -f "dangling=true" -q)
"docker rmi" requires at least 1 argument.
See 'docker rmi --help'.

Usage:  docker rmi [OPTIONS] IMAGE [IMAGE...]

Remove one or more images
builder /var/lib/docker # docker volume rm $(docker volume ls -qf dangling=true)
"docker volume rm" requires at least 1 argument.
See 'docker volume rm --help'.

Usage:  docker volume rm [OPTIONS] VOLUME [VOLUME...]

Remove one or more volumes
builder /var/lib/docker # docker ps -s                            
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES               SIZE
```

It seems the issue is that all containers are rightfully gone, but the space remains allocated, until I reboot and nuke everything. The [accepted answer](https://stackoverflow.com/a/27894675/1893275) in the aforementioned thread also happens to be a nuclear one. The problem with that, though, is that stopping and starting the daemon alone won't allow us to delete the devicemapper volume, we also have to reboot... 

It might be, though, that we just need to set up a more sane docker configuration (as intimated [here](https://docs.docker.com/storage/storagedriver/device-mapper-driver/#configure-direct-lvm-mode-for-production)) - since apparently [Docker's default is to be broken](https://github.com/moby/moby/issues/3182#issuecomment-196269192)

In fact it seems we are the beneficiaries of this docker design choice (note the last line)
```
builder /var/lib/docker # docker info
Containers: 0                                                                               
 Running: 0                                                                          
 Paused: 0                                                        
 Stopped: 0                                                                                                                                     
Images: 5                     
Server Version: 17.09.1-ce
Storage Driver: devicemapper
 Pool Name: docker-0:19-7207861-pool
 Pool Blocksize: 65.54kB
 Base Device Size: 10.74GB
 Backing Filesystem: ext4                            
 Data file: /dev/loop0   
 Metadata file: /dev/loop1                                                                                                 
 Data Space Used: 9.259GB                     
 Data Space Total: 107.4GB  
 Data Space Available: 57GB
 Metadata Space Used: 5.19MB                         
 Metadata Space Total: 2.147GB
 Metadata Space Available: 2.142GB            
 Thin Pool Minimum Free Space: 10.74GB                                    
 Udev Sync Supported: true               
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0                   
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.119 (2016-03-04)                                    
Logging Driver: json-file                 
Cgroup Driver: cgroupfs 
Plugins:
 Volume: local                               
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive                                                                 
Runtimes: runc                                  
Default Runtime: runc         
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84 (expected: 3f2f8b84a77f73d38244dd690525642a72156c64)
init version: v0.16.1 (expected: 949e6facb77383876aeff8a6944dde66b3089574)
Security Options:                     
 seccomp                                                                                                                                        
  Profile: default                                
Kernel Version: 4.9.76-gentoo-r1                                                                                                                                                                                                  Operating System: Gentoo/Linux                
OSType: linux                                        
Architecture: x86_64                                 
CPUs: 8
Total Memory: 30.64GiB
Name: builder
ID: YWRP:5QVL:X5JB:ZDEK:H643:O74Y:CYLY:CQ2S:26AW:EZZ6:5HVM:6NPE
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: chymera
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: devicemapper: usage of loopback devices is strongly discouraged for production use.
         Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
```

The problem is, I have no idea what block device it would make sense to select :-/
